docs(client): add Java client docs

vcrfxia opened the pull request:
### Description 

Draft of docs for new Java client. @MichaelDrogalis had some ideas for integrating these with the new use case tutorials, which is a potential direction for these docs to evolve.

API docs generated from javadocs on the interfaces are WIP, and will likely be a separate PR.

### Testing done 

Docs-only change.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

MichaelDrogalis commented:
Great start @vcrfxia! Something that jumped out at me while thinking through the doc structure is what this will look like as we add new clients. One thing we'll certainly need is the ability to directly link to a specific language's doc set. To that end, I think it would be organizationally tidy if we tuck all of each client's details into a single page and nest it under a top-level "Clients" section in the reference. That makes finding the whole set of clients that we offer simple, and each is directly linkable.

One other thought that I had is that we ought to think through the top-level headings for each client. I realize we're still working on the functionality, but the key things to me when I'm using it are: how do I install it, how do I start some persistent queries, how do I feed in some data, and how do I run push and pull queries? There's probably a common set of things to rally around each time we document a client (plus some inevitable miscellaneous stuff that's parochial to each language).

I'm still mulling over how we could use tabbed sets of code blocks to compare and contrast languages. It seems like there's something cool here, but we should figure out how to make each language reference findable and linkable.

colinhicks commented:
> API docs generated from javadocs on the interfaces are WIP, and will likely be a separate PR.

@vcrfxia let's catch up about the tooling side of this. Considering future support for APIs in other languages, we might want to use something other than the Maven Javadoc plugin for example. We'd still use javadocs as in the syntax, though.

> it would be organizationally tidy if we tuck all of each client's details into a single page and nest it under a top-level "Clients" section in the reference.

+1 for a top-level Clients section. @MichaelDrogalis to clarify, you mean that each client would have its own page within this section, rather than a single page with all the details for every client, correct?

In my opinion we should use tabbed code blocks sparingly, if at all. They present usability issues and would be harder to maintain.

MichaelDrogalis commented:
@colinhicks Yeah, that is what I meant.

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+title: Java Client for ksqlDB
+tagline: Java client for ksqlDB
+description: Send requests to ksqlDB from your Java app
+---
```
```suggestion
keywords: ksqlDB, java, client
---
```

JimGalasyn replied:
And other relevant keywords.

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Receive query results one row at a time (streamQuery())<a name="stream-query"></a>
+----------------------------------------------------------------------------------
+
+The `streamQuery()` method allows client apps to receive query results one row at a time,
```
```suggestion
The `streamQuery()` method enables client apps to receive query results one row at a time,
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+}
+```
+
+This method may be used to issue both push and pull queries, though the usage pattern is most suited for push queries.
````
```suggestion
You can use this method to issue both push and pull queries, but the usage pattern is better for push queries.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
+
+By default, push queries return results starting from the beginning of the stream or table.
+To instead start from the end and only receive newly arriving rows, set the property `auto.offset.reset` to `latest`.
```
```suggestion
To start from the end and receive only newly arriving rows, set the `auto.offset.reset` property to `latest`.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+
+### Asynchronous Usage ###
+
+To consume records in an asynchronous fashion, first create a Reactive Streams subscriber to receive query result rows:
```
```suggestion
To consume records asynchronously, create a [Reactive Streams](http://www.reactive-streams.org/) subscriber to receive query result rows:
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+}
+```
+
+Then, use the client to send the query result to the server and stream results to the subscriber:
````
```suggestion
Use the Java client to send the query result to the server and stream results to the subscriber:
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+### Synchronous Usage ###
+
+To consume records one-at-a-time in a synchronous fashion, use the `poll()` method on the query result object.
+If `poll()` is called with no arguments, `poll()` will block until a new row becomes available or the query is terminated.
```
```suggestion
If `poll()` is called with no arguments, `poll()` blocks until a new row becomes available or the query is terminated.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+
+To consume records one-at-a-time in a synchronous fashion, use the `poll()` method on the query result object.
+If `poll()` is called with no arguments, `poll()` will block until a new row becomes available or the query is terminated.
+You can also pass a `Duration` argument to `poll()` which will cause `poll()` to return null if no new rows are received by the time the duration has elapsed.
```
```suggestion
You can also pass a `Duration` argument to `poll()`, which causes `poll()` to return `null` if no new rows are received by the time the duration has elapsed.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+To consume records one-at-a-time in a synchronous fashion, use the `poll()` method on the query result object.
+If `poll()` is called with no arguments, `poll()` will block until a new row becomes available or the query is terminated.
+You can also pass a `Duration` argument to `poll()` which will cause `poll()` to return null if no new rows are received by the time the duration has elapsed.
+See [the API reference](TODO) for more.
```
```suggestion
For more information, see the [API reference](TODO).
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Receive query results in a single batch (executeQuery())<a name="execute-query"></a>
+------------------------------------------------------------------------------------
+
+The `executeQuery()` method allows client apps to receive query results as a single batch,
```
```suggestion
The `executeQuery()` method enables client apps to receive query results as a single batch,
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+------------------------------------------------------------------------------------
+
+The `executeQuery()` method allows client apps to receive query results as a single batch,
+returned once the query has completed.
```
```suggestion
that's returned when the query completes.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+}
+```
+
+This method is suitable for both pull queries as well as terminating push queries (e.g., those with a `LIMIT` clause).
````
```suggestion
This method is suitable for both pull queries and terminating push queries, for example, queries that have a `LIMIT` clause.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+This method is suitable for both pull queries as well as terminating push queries (e.g., those with a `LIMIT` clause).
+For non-temrinating push queries, use [the `streamQuery()` method](./stream-query.md) instead. 
+
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
```
```suggestion
Query properties can be passed as an optional second argument. For more information, see the [client API reference](TODO).
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
+
+By default, push queries return results starting from the beginning of the stream or table.
+To instead start from the end and only receive newly arriving rows, set the property `auto.offset.reset` to `latest`.
```
```suggestion
To start from the end and receive only newly arriving rows, set the `auto.offset.reset` property to `latest`.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Terminate a push query (terminatePushQuery())<a name="terminate-push-query"></a>
+--------------------------------------------------------------------------------
+
+The `terminatePushQuery()` method allows client apps to terminate push queries.
```
```suggestion
The `terminatePushQuery()` method enables client apps to terminate push queries.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+}
+```
+
+The query ID is obtained from the query result response object when push queries are issued via the client,
````
```suggestion
The query ID is obtained from the query result response object when the client issues push queries,
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+```
+
+The query ID is obtained from the query result response object when push queries are issued via the client,
+via either [`streamQuery()`](./stream-query.md) or [`executeQuery()`](./execute-query.md).
````
```suggestion
by using either the [`streamQuery()`](./stream-query.md) or [`executeQuery()`](./execute-query.md) methods.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+
+### Example Usage ###
+
+Here's an example of terminating a push query issued via `streamQuery()`:
```
```suggestion
Here's an example of terminating a push query issued by using the `streamQuery()` method:
```

vcrfxia replied:
Out of curiosity, why the preference for "a push query issued by using the `streamQuery()` method" rather than "a push query issued using the `streamQuery()` method"? I've currently updated to the latter but can easily switch to your original suggestion instead. (Can't quite pin it but the extra "by" reads weirdly to me, haha)

purplefox replied:
I think strictly speaking (grammatically) Jim is right... (?)

I.e. it seems more natural for me to say "a push query issued _by_ using the.." or "a push query issued _while_ using the..." It seems to me it needs some word in there.

Having said that.... it probably doesn't matter a great deal ;)

https://forum.wordreference.com/threads/relative-pronouns-omitting-relative-pronoun-not-relative-adverb.282370/

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
````diff
@@ -0,0 +1,416 @@
+}
+```
+
+And here's an analogous example for terminating a push query issued via `executeQuery()`:
````
```suggestion
And here's an analogous example for terminating a push query issued by using the `executeQuery()` method:
```

vcrfxia replied:
Ditto

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,416 @@
+Insert a batch of rows into a stream (insertInto())<a name="insert-into"></a>
+-----------------------------------------------------------------------------
+
+Client apps can insert rows of data into existing ksqlDB streams via the `insertInto()` method.
```
```suggestion
Client apps can insert rows of data into existing ksqlDB streams by using the `insertInto()` method.
```

JimGalasyn commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+This method may be used to issue both push and pull queries, though the usage pattern is most suited for push queries.
+For pull queries, consider [the `executeQuery()` method](./execute-query.md) instead. 
+
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
```
```suggestion
Query properties can be passed as an optional second argument. For more information, see the [client API reference](TODO).
```

vcrfxia commented:
Thanks for the review @JimGalasyn ! Applied your suggestions.

> It's a good idea to add an index.md so customers won't 404 when they go to https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-clients/. Also, we should have an entry in developer-guide/index.md.

Done and done. developer-guide/index.md currently links to the new index page which is a bit sparse. Let me know if you think that's a problem. Thanks!

vcrfxia commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+  @Override
+  public synchronized void onSubscribe(final Subscription subscription) {
+    System.out.println("Subscriber is subscribed.");
+    this.subscription = subscription;
```
Had it below but it makes more sense to have it here. Good call.

vcrfxia commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+  @Override
+  public synchronized void onNext(final Row row) {
+    System.out.println("Received a row!");
+    System.out.println("Row: " + row.values());
```
> We can consider providing a simple Subscriber class (like our BaseSubscriber) that does the requests for you to make things a bit simpler.

This makes sense. I was thinking we'd want to provide something similar for the Publisher in order to make the streaming INSERT INTO method more easily usable. Do you think it's critical this makes it into 0.10.0 or are we okay with it in 0.11.0 instead?

vcrfxia replied:
Would we want to include the same Vert.x context checking logic in our BaseSubscriber in the one provided by the client?

purplefox replied:
I think 0.11.0 is ok.
I don't think we'd want the context checking as we can't be sure what threads the user would use the subscriber from.

vcrfxia commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+final StreamedQueryResult streamedQueryResult;
+try {
+  streamedQueryResult = client.streamQuery("SELECT * FROM MY_STREAM EMIT CHANGES;").get();
+} catch (Exception e) {
```
I agree it clutters the example but I think it's useful to have a fully fleshed out example that users can copy-paste from. @colinhicks and @MichaelDrogalis mentioned adding to the use case tutorials with examples of how to run tutorial commands from the client instead of the API. Maybe once we do that we can remove the try-catch's here?

Would be curious to hear what @JimGalasyn understands is standard for docs like these -- do we value readability over copy-paste-ability?

purplefox replied:
My 2c: Docs should always be optimised for clarity and readibility and we should avoid anything that detracts from the intention of the example. Otherwise we risk turning off users and giving the impression the API is harder to use than it really is.
Even without try...catch the example is copy and pastable - any Java dev knows they need to handle any checked exceptions.

purplefox replied:
Also let's lose the finals please :)

vcrfxia commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+final List<Row> resultRows;
+try {
+  resultRows = batchedQueryResult.get();
+} catch (Exception e) {
```
As above.

vcrfxia commented on docs/developer-guide/ksqldb-clients/java-client.md:
```diff
@@ -0,0 +1,387 @@
+    .put("USER_ID", "User_321"));
+
+try {
+  client.insertInto("ORDERS", row).get();
```
What kind of composition did you have in mind? Composing multiple insertInto() requests, or something else?

purplefox replied:
Could be anything really - e..g. executing a query then inserting some rows into a stream for each result, or inserting into a database.

vcrfxia commented:
Hey @MichaelDrogalis the client code on master is ready for a test drive. I'm still working on adding more test coverage and squashing bugs as they come up but the functionality's all there :)

purplefox commented:
I think this is a good start, and it answers of the question of _how_ to use the client, but I think we also need to provide something that answers the question of _what_ it is and _why_ someone would want to use the client.

I'd like to see an introductory section that contains something like:

What is it?

ksqlDB ships with a Java client which enables you to.... blah blah. It currently supports queries and inserts but admin operations will be added before long.... blah blah The client uses the HTTP2 API... blah blah... The Java client is designed for use in Java applications that blah.. blah.. It allows you to execute operations against ksqlDB without manually crafting HTTP requests (although you can do that yourself if you want)... blah blah

I know sometimes I get frustrated when I heard about new "project X" then I google it and all I get is some level API docs, but nothing that actually tells what project X _is_ ;)

vcrfxia commented:
Thanks for the review @purplefox . Made the following changes:

- Removed all instances of try-catch and final
- Added an intro blurb at the top
- Added more complex examples at the bottom showcasing future composition. (The examples draw from the use-case tutorials, an idea @MichaelDrogalis tossed around.)

vcrfxia commented:
Applied the latest feedback from https://github.com/confluentinc/ksql/pull/5581 and retargeted to the 6.0.x branch. Will merge once the build passes.
